File: python/scextractor.py
import re

class ShortcodeExtractor:    
    def __init__(self):
        self.text = ""
        self.bracket_type = "curly"
        self.bracket_open = "{"
        self.bracket_close = "}"
        self.variables = {}
        self.regex_pattern = ""

    def set_text(self, text):
        self.text = text
        return self

    def set_bracket_type(self, type):
        self.bracket_type = type
        return self

    def set_variables(self, variables):
        self.variables = variables
        return self

    def extract(self):
        self.regex_build()
        regex_variables = re.findall(self.regex_pattern, self.text)
        new_text = self.text
        for va in regex_variables:
            val = self.variables.get(va)
            if val:
                new_text = new_text.replace(str(self.bracket_open) + str(va) + str(self.bracket_close), val)

        return new_text
    
    def regex_build(self):
         match self.bracket_type:
            case 'curly':
                self.regex_pattern = r'{({*[^{}]*}*)}'
                self.bracket_open = "{"
                self.bracket_close = "}"
            case 'square':
                self.regex_pattern = r'\[([^\]\[\r\n]*)\]'
                self.bracket_open = "["
                self.bracket_close = "]"
            case 'parentheses':
                self.regex_pattern = r'\(([^\)]+)\)'
                self.bracket_open = "("
                self.bracket_close = ")"
            case _:
                print("Invalid bracket type")
                exit()

File: golang/scextractor.go
package scextractor

import (
	"regexp"
	"strings"
)

type BracketType string

const (
	Curly       BracketType = "curly"
	Square      BracketType = "square"
	Parentheses BracketType = "parentheses"
)

var text string
var bracketType BracketType
var bracketOpen string
var bracketClose string
var regexPattern string
var variables map[string]interface{}

func main() {
	bracketType = "curly"
}

func SetText(txt string) {
	text = txt
}

func SetBracketType(brcktType BracketType) {
	bracketType = brcktType
}

func SetVariables(vars map[string]interface{}) {
	variables = vars
}

func Extract() string {
	regexBuild()
	r := regexp.MustCompile(regexPattern)
	matches := r.FindAllString(text, -1)

	re := regexp.MustCompile(`[` + bracketOpen + bracketClose + `]`)
	newText := text
	for _, m := range matches {
		key := re.ReplaceAllString(m, "")
		val := variables[key]

		if val != nil {
			newText = strings.Replace(newText, m, val.(string), 1)
		}
	}

	return newText
}

func regexBuild() {
	switch bracketType {
	case "curly":
		regexPattern = `{({*[^{}]*}*)}`
		bracketOpen = "{"
		bracketClose = "}"
	case "square":
		regexPattern = `\[([^\]\[\r\n]*)\]`
		bracketOpen = `\[`
		bracketClose = `\]`
	case "parentheses":
		regexPattern = `\(([^\)]+)\)`
		bracketOpen = "("
		bracketClose = ")"
	default:
		panic("Invalid bracket type")
	}
}
